src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs

```diff
@@ -0,0 +1,308 @@
+        /// at the same time.
+        /// </remarks>
+        private T _item0;
+
+#pragma warning disable IDE0044 // Add readonly modifier
+#pragma warning disable CS0649 // Field 'name' is never assigned to, and will always have its default value
+        /// <summary>
+        /// The second inline element.
+        /// </summary>
+        /// <remarks>
+        /// This field is accessed by offset from <see cref="_item0"/>. See <see cref="_item0"/> for usage information.
+        /// </remarks>
+        private T _item1;
+
+        /// <summary>
+        /// The third inline element.
+        /// </summary>
+        /// <remarks>
+        /// This field is accessed by offset from <see cref="_item0"/>. See <see cref="_item0"/> for usage information.
+        /// </remarks>
+        private T _item2;
+
+        /// <summary>
+        /// The fourth inline element.
+        /// </summary>
+        /// <remarks>
+        /// This field is accessed by offset from <see cref="_item0"/>. See <see cref="_item0"/> for usage information.
+        /// </remarks>
+        private T _item3;
+#pragma warning restore CS0649 // Field 'name' is never assigned to, and will always have its default value
+#pragma warning restore IDE0044 // Add readonly modifier
+
+        /// <summary>
+        /// The number of inline elements held in the array. This value is only used when <see cref="_builder"/> is
+        /// <see langword="null"/>.
+        /// </summary>
+        private int _count;
+
+        /// <summary>
+        /// A builder used for dynamic storage of collections that may exceed the limit for inline elements.
+        /// </summary>
+        /// <remarks>
+        /// This field is initialized to non-<see langword="null"/> the first time the <see cref="TemporaryArray{T}"/>
+        /// needs to store more than four elements. From that point, <see cref="_builder"/> is used instead of inline
+        /// elements, even if items are removed to make the result smaller than four elements.
+        /// </remarks>
+        private ArrayBuilder<T>? _builder;
+
+        private TemporaryArray(in TemporaryArray<T> array)
+        {
+            // Intentional copy used for creating an enumerator
+#pragma warning disable RS0042 // Do not copy value
+            this = array;
+#pragma warning restore RS0042 // Do not copy value
+        }
+
+        public static TemporaryArray<T> Empty => default;
+
+        public readonly int Count => _builder?.Count ?? _count;
+
+        public T this[int index]
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            readonly get
+            {
+                if (_builder is not null)
+                    return _builder[index];
+
+                if ((uint)index >= _count)
+                    ThrowIndexOutOfRangeException();
+
+                return Unsafe.Add(ref Unsafe.AsRef(in _item0), index);
```

Comment:
are all these ops safe for any sort of T?   your tests all tested unmanaged types.  I admit i'm a bit wary about this for normal managed ref types.

Reply:
I can add an assertion to a static constructor that `Unsafe.Add(in _item0, 1)` gives the address of `_item1`.

@tannergooding is the one who originally pointed out this form.

Reply:
Its safe for blittable data. I don't believe the VM will reorder for reference types either when all fields are T, but someone on the runtime team can likely confirm.

@jkotas?

Reply:
If you go by the spec, the problem with this is that the may choose to enregister everything (AggressiveInlining on the method helps with that), it won't see any references of `_item1`, `_item2`, ... , and drops the storage for them.

The JIT has compatibility quirks to make these sort of cases work because of you are not the first one using this hack. I do not know the exact conditions when these compatibility quirks kick in.

You can make this more reliable by marking these methods non-inlineable.

Reply:
@jkotas I have GC.KeepAlive references to all four fields in the Dispose method, which should directly address this

Reply:
I'll switch to a switch expression for now since we need the rest of this win, and we can come back to this separately

Reply:
> I have GC.KeepAlive references to all four fields in the Dispose method, which should directly address this

It is not enough to guarantee that the values are going to be spilled onto stack sequentially.

Reply:
@sharwell  can we not just switch on the _count and get/set the appropriate item?  i'm struggling to think of a case where that would actuallybe a problem in the ide scenarios.

Reply:
> can we not just switch on the _count and get/set the appropriate item?

Already done in accordance with https://github.com/dotnet/roslyn/pull/48896#discussion_r513073184

> i'm struggling to think of a case where that would actuallybe a problem in the ide scenarios.

The difference was _obvious_ in the benchmarks for https://github.com/tunnelvisionlabs/dotnet-trees/pull/102.